File: im-mobile/lib/services/friend_recommendation_service.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import 'package:im_mobile/config/api_config.dart';
import 'package:im_mobile/models/recommended_user_model.dart';
import 'package:im_mobile/utils/http_util.dart';

class FriendRecommendationService {
  static final FriendRecommendationService _instance = FriendRecommendationService._internal();
  factory FriendRecommendationService() => _instance;
  FriendRecommendationService._internal();

  final Dio _dio = HttpUtil().dio;

  Future<List<RecommendedUserModel>> getRecommendations({
    required int page,
    required int size,
    String? algorithm,
  }) async {
    try {
      final response = await _dio.get(
        '${ApiConfig.baseUrl}/friends/recommendations',
        queryParameters: {
          'page': page,
          'size': size,
          if (algorithm != null && algorithm != 'mixed') 'algorithm': algorithm,
        },
      );

      if (response.statusCode == 200 && response.data != null) {
        final data = response.data['data'] as List<dynamic>?;
        if (data != null) {
          return data.map((json) => RecommendedUserModel.fromJson(json)).toList();
        }
      }
      return [];
    } catch (e) {
      throw Exception('Failed to load recommendations: $e');
    }
  }

  Future<void> sendFriendRequest(String userId) async {
    try {
      final response = await _dio.post(
        '${ApiConfig.baseUrl}/friends/requests',
        data: {
          'toUserId': userId,
          'source': 'recommendation',
        },
      );

      if (response.statusCode != 200 && response.statusCode != 201) {
        throw Exception('Failed to send friend request');
      }
    } catch (e) {
      throw Exception('Failed to send friend request: $e');
    }
  }

  Future<void> ignoreRecommendation(String userId) async {
    try {
      final response = await _dio.post(
        '${ApiConfig.baseUrl}/friends/recommendations/$userId/ignore',
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to ignore recommendation');
      }
    } catch (e) {
      throw Exception('Failed to ignore recommendation: $e');
    }
  }

  Future<Map<String, dynamic>> getRecommendationStats() async {
    try {
      final response = await _dio.get(
        '${ApiConfig.baseUrl}/friends/recommendations/stats',
      );

      if (response.statusCode == 200 && response.data != null) {
        return response.data;
      }
      return {};
    } catch (e) {
      throw Exception('Failed to load stats: $e');
    }
  }

  Future<List<RecommendedUserModel>> refreshRecommendations() async {
    try {
      final response = await _dio.post(
        '${ApiConfig.baseUrl}/friends/recommendations/refresh',
      );

      if (response.statusCode == 200 && response.data != null) {
        final data = response.data['data'] as List<dynamic>?;
        if (data != null) {
          return data.map((json) => RecommendedUserModel.fromJson(json)).toList();
        }
      }
      return [];
    } catch (e) {
      throw Exception('Failed to refresh recommendations: $e');
    }
  }
}

File: im-desktop/src/services/friendRecommendationService.ts
import { apiClient } from '../utils/apiClient';

export interface RecommendationRequest {
  page: number;
  size: number;
  algorithm?: 'mutual_friends' | 'interest_tags' | 'group_relation' | 'mixed';
}

export interface RecommendationResponse {
  userId: string;
  nickname: string;
  avatar: string;
  reasonType: 'mutual_friends' | 'interest_tags' | 'group_relation' | 'mixed';
  reasonDescription: string;
  score: number;
  mutualFriendsCount?: number;
  commonTags?: string[];
  commonGroups?: string[];
}

export interface RecommendationResult {
  data: RecommendationResponse[];
  total: number;
  page: number;
  size: number;
}

export interface RecommendationStats {
  totalRecommended: number;
  totalIgnored: number;
  totalAccepted: number;
  acceptanceRate: number;
}

class FriendRecommendationService {
  private readonly baseUrl = '/api/v1/friends/recommendations';

  async getRecommendations(request: RecommendationRequest): Promise<RecommendationResult> {
    const params = new URLSearchParams();
    params.append('page', request.page.toString());
    params.append('size', request.size.toString());
    if (request.algorithm && request.algorithm !== 'mixed') {
      params.append('algorithm', request.algorithm);
    }

    const response = await apiClient.get<RecommendationResult>(
      `${this.baseUrl}?${params.toString()}`
    );
    return response.data;
  }

  async sendFriendRequest(userId: string): Promise<void> {
    await apiClient.post('/api/v1/friends/requests', {
      toUserId: userId,
      source: 'recommendation'
    });
  }

  async ignoreRecommendation(userId: string): Promise<void> {
    await apiClient.post(`${this.baseUrl}/${userId}/ignore`);
  }

  async getRecommendationStats(): Promise<RecommendationStats> {
    const response = await apiClient.get<RecommendationStats>(`${this.baseUrl}/stats`);
    return response.data;
  }

  async refreshRecommendations(): Promise<RecommendationResult> {
    const response = await apiClient.post<RecommendationResult>(`${this.baseUrl}/refresh`);
    return response.data;
  }
}

export const friendRecommendationService = new FriendRecommendationService();
